ykrt/src/compile/jitc_yk/opt/lvn.rs

```diff
@@ -0,0 +1,115 @@
+#[cfg(test)]
+use crate::compile::jitc_yk::jit_ir::ConsumeInst;
+use crate::compile::{
+    jitc_yk::jit_ir::{AddInst, Inst, InstIdx, Module, Operand},
+    CompilationError,
+};
+use typed_index_collections::TiVec;
+
+/// Perform (very very) simple dead-code elimination.
+pub(super) fn lvn(mut m: Module) -> Result<Module, CompilationError> {
+    // Which operations are used?
+    let mut used_ops = vec![false; m.len()];
+    for inst in m.iter_mut_insts() {
```

Comment:
This looks like a def/use analysis.

We could generalise and abstract this into an API and then it could be shared with the register allocator, which will also need to know this kind of stuff (e.g. a local is born upon its definition, a local dies after it's last use).

It's a shame we will have to manually write the code to find operands in instructions. I guess that was one advantage of the "dynamic" IR: the operands were just there is a list. I'm not suggesting we go back to a dynamic IR though!

Reply:
> This looks like a def/use analysis.

Mostly, but this code is not meant to be a usable optimisation pass: it's meant to show us how *an* optimisation pass looks in our code. The analysis is incomplete beyond the very simple example.

Reply:
I guess we could just define a function `operands()` on each instructions which returns all operands?

Reply:
That's definitely one possibility. Honestly, I'm currently unsure what the best approach might be.